File: python/psfs/psf_vectorial.py

# import time
import os
import pathlib
from dataclasses import dataclass

# import tensorflow as tf
import numpy as np
from scipy.interpolate import RegularGridInterpolator

from .psf import Psf
from .library import effField
from .library import effIntensity
from .library import conv2fft
from .library import convnfft

@dataclass
class PSFStruct:
    normalization : float = 1
    interp : callable = None

def addzernikeaberrations(sys, addpar):
    if 'Zr' in sys:
        sys['Ei'].append('zernike')
    elif 'Zr' in addpar:
        sys['Ei'].append('zernike')
        sys['Zr'] = addpar['Zr']
    return sys

def normpsf(psf):
    return psf, np.max(psf)

def range4PSF(psf,dr,dz):
    nx = (np.arange(psf.shape[0]) - (psf.shape[0]-1)/2)*dr*1e9
    nz = (np.arange(psf.shape[2]) - (psf.shape[2]-1)/2)*dz*1e9
    
    return nx, nx, nz

def meshgrid4PSF(psf,dr,dz):
    nx, nz = range4PSF(psf,dr,dz)
    X, Y, Z = np.meshgrid(nx, nx, nz)
    
    return X, Y, Z

def genkey(phasepattern, Lxs=None):
    if Lxs is None:
        return f"{phasepattern}"
    elif isinstance(Lxs, (int, float)):
        Lx = [[Lxs]]
    else:
        # list or array (we hope)
        Lx = list(Lxs.copy())
        for i, x in enumerate(Lxs):
            if not isinstance(x , list):
                Lx[i] = [x]

    # unlist
    return f"{phasepattern}" + "".join([str(x) for xs in Lx for x in xs])

class PsfVectorial(Psf):
    """
    PSF class that uses a 3D volume to describe the PSF.
    Should only be used with single-channel data.
    """
    def __init__(self, parameters=None) -> None:
        super().__init__()
        
        if parameters is None:
            import yaml

            base_path = pathlib.Path(__file__).parent.parent
            param_path = os.path.join(base_path, 'settings/default_microscope.yaml')

            with open(param_path, 'r') as fn:
                self.parameters = yaml.load(fn, Loader=yaml.FullLoader)
        else:
            self.parameters = parameters
        self.PSFs = {}
        self.PSFph = None
        self.normfactgauss = None
        self.pinholepar = None
        self.sigma = 310/2.355
        self.fwhm = 310

    def setpar(self, **kwargs):
        # # overwrite any of the parameter fields if they exist
        # intersection_keys = list(self.parameters.keys() & kwargs.keys())

        # for key in intersection_keys:
        #     self.parameters[key].update(kwargs[key])

        # # otherwise add to addpar
        # exclusive_keys = list(set(kwargs.keys()) - set(self.parameters.keys()))
        # if len(exclusive_keys) > 0:
        #     if 'addpar' not in self.parameters.keys():
        #         self.parameters['addpar'] = {}
        #     for key in exclusive_keys:
        #         self.parameters['addpar'][key] = kwargs[key]

        # if len(intersection_keys) > 0:
        #     # recalculate
        #     self.PSFs = {}
        #     self.normfactgauss = None
        #     if self.pinholepar is not None:
        #         self.setpinhole(**self.pinholepar)
        fn = kwargs.keys()
        fnpar = self.parameters.keys()

        recalculate = False
        for k in fn:
            assigned = False
            for j in fnpar:
                if k in self.parameters[j].keys():
                    try:
                        if self.parameters[j][k] != kwargs[k]:
                            self.parameters[j][k] = kwargs[k]
                            recalculate = True
                    except ValueError:
                        # probably an array
                        if np.any(self.parameters[j][k] != kwargs[k]):
                            self.parameters[j][k] = kwargs[k]
                            recalculate = True
                    assigned = True
            if not assigned:
                self.parameters['addpar'][k] = kwargs[k]

        if recalculate:
            self.PSFs = {}
            self.normfactgauss = None
            if self.pinholepar is not None:
                self.setpinhole(**self.pinholepar)

    def intensity(self, flpos, patternpos, phasepattern, L=None):
        key = genkey(phasepattern, L)

        try:
            psfint = self.PSFs[key]
        except KeyError:
            key = self.calculatePSFs(phasepattern, L)
            psfint = self.PSFs[key]

        flposrel = flpos - patternpos
        iexc = psfint.interp(flposrel) + self.zerooffset
        iexc = np.maximum(iexc,0)
        try:
            phfac = self.PSFs[self.PSFph].interp(flpos)
        except KeyError:
            phfac = np.ones_like(iexc)
        idet = iexc * phfac

        return idet, phfac
    
    def calculatePSFs(self, phasepattern, Lxs=None, forcecalculation=False):
        key = genkey(phasepattern, Lxs)

        if key in self.PSFs and not forcecalculation:
            return key
        
        if Lxs is None:
            # safety, but not so safe so alert the user
            Lxs = [0]
            print("Warning! Unknown Lxs. Using a default of [0].")
        
        intmethod = 'cubic' # linear,cubic?
        bounds_error = False
        extraolation_method = None  # If None, extrapolate
        
        print(f"{key}, ", end="")

        opt = self.parameters['opt']
        out = self.parameters['addpar']
        sys = self.parameters['sys']

        fwhm = 0.51*sys['lo']/sys['NA']*1e9
        self.sigma = fwhm/2.35

        if self.normfactgauss is None:
            sys['Ei'] = ['circular']
            outc = effField(sys, out, opt)
            outc = effIntensity(sys, outc)
            zmid=int(np.ceil(outc['I'].shape[2]/2))
            self.normfactgauss=np.max(outc['I'][:,:,zmid])  # normalized to max =1

        if phasepattern == "flat":
            sys['Ei'] = ['circular']
            sys = addzernikeaberrations(sys,out)
            out = effField(sys, out, opt)
            out = effIntensity(sys,out)
            PSF = out['I'].squeeze()/self.normfactgauss
            PSF = self.beadsize(PSF, sys['beadradius'])
            PSFdonut = PSFStruct()
            PSF, PSFdonut.normalization = normpsf(PSF)
            xx, yy, zz = range4PSF(PSF,out['dr'],out['dz'])
            PSFdonut.interp = RegularGridInterpolator((xx, yy, zz), 
                                                       PSF, 
                                                       method = intmethod, 
                                                       bounds_error = bounds_error,
                                                       fill_value = extraolation_method)
            self.PSFs[key] = PSFdonut
        elif "halfmoon" in phasepattern:
            # convert L into phase
            dxdphi = 1.4  # nm/degree, from LSA paper Deguchi
            # dzdphi = -3.6  #nm/degree
            sys['delshift'] = np.deg2rad(float(Lxs)/dxdphi)
            sys['Ei'] = ['halfmoon', 'linear']
            sys = addzernikeaberrations(sys, out)
            out = effField(sys, out, opt)
            out = effIntensity(sys, out)
            PSF = out['I'].squeeze()/self.normfactgauss
            PSF = self.beadsize(PSF, sys['beadradius'])
            PSFx = PSFStruct()
            PSFy = PSFStruct()
            PSF, PSFx.normalization = normpsf(PSF)
            xx, yy, zz = range4PSF(PSF,out['dr'],out['dz'])
            PSFx.interp = RegularGridInterpolator((xx, yy, zz), 
                                                   PSF.transpose(1,0,2), 
                                                   method = intmethod, 
                                                   bounds_error = bounds_error,
                                                   fill_value = extraolation_method)
            PSFy.interp = RegularGridInterpolator((xx, yy, zz), 
                                                   PSF, 
                                                   method = intmethod, 
                                                   bounds_error = bounds_error,
                                                   fill_value = extraolation_method)   
            PSFy.normalization = PSFx.normalization
            self.PSFs[genkey("halfmoonx",Lxs)] = PSFx
            self.PSFs[genkey("halfmoony",Lxs)] = PSFy
        elif phasepattern == "pinhole":
            sys['Ei'] = ['circular']
            sys = addzernikeaberrations(sys, out)
            phdiameter = float(Lxs[0])
            if len(Lxs) == 3:
                phpos = [float(Lxs[1]), float(Lxs[2])]
            else:
                phpos = [0, 0]
            out = effField(sys, out, opt)
            out = effIntensity(sys, out)
            psfg = out['I'].squeeze()
            norm = psfg[:,:,int(round(psfg.shape[2]/2))].sum()
            psfg /= norm
            pixelsize = opt['pixSize']*1e9
            nx = np.arange(psfg.shape[0])
            nx = (nx-np.mean(nx))*pixelsize
            Xk, Yk = np.meshgrid(nx, nx) 
            kernel = np.double(((Xk - phpos[1])**2 + (Yk - phpos[0])**2) < (phdiameter / 2)**2)
            # print(kernel.shape, psfg.shape)
            psfph = conv2fft(psfg, kernel)
            xx, yy, zz = range4PSF(psfph, out['dr'], out['dz'])
            PSFdonut = PSFStruct()
            PSFdonut.interp = RegularGridInterpolator((xx, yy, zz),
                                                      psfph, 
                                                      method = intmethod, 
                                                      bounds_error = bounds_error,
                                                      fill_value = extraolation_method)
            self.PSFs[key] = PSFdonut
        elif phasepattern == "tophat":
            dzdphi = -3.6  #nm/degree
            sys['delshift'] = np.deg2rad(float(Lxs)/dzdphi)
            sys['Ei'] = ['pishift', 'circular']
            sys = addzernikeaberrations(sys, out)
            out = effField(sys, out, opt)
            out = effIntensity(sys, out)
            PSF = out['I'].squeeze()/self.normfactgauss
            PSF = self.beadsize(PSF, sys['beadradius'])
            PSFdonut = PSFStruct()
            PSF, PSFdonut.normalization = normpsf(PSF)
            xx, yy, zz = range4PSF(PSF,out['dr'],out['dz'])
            PSFdonut.interp = RegularGridInterpolator((xx, yy, zz), 
                                                       PSF, 
                                                       method = intmethod, 
                                                       bounds_error = bounds_error,
                                                       fill_value = extraolation_method)
            self.PSFs[key] = PSFdonut
        elif phasepattern == "vortex":
            sys['Ei'] = ['phaseramp', 'circular']
            sys = addzernikeaberrations(sys, out)
            out = effField(sys, out, opt)
            out = effIntensity(sys, out)
            PSF = out['I'].squeeze()/self.normfactgauss
            PSF = self.beadsize(PSF, sys['beadradius'])
            PSFdonut = PSFStruct()
            PSF, PSFdonut.normalization = normpsf(PSF)
            xx, yy, zz = range4PSF(PSF, out['dr'], out['dz'])
            # print(xx.shape, yy.shape, zz.shape, PSF.shape, PSF.dtype, np.sum(np.isnan(PSF)))
            PSFdonut.interp = RegularGridInterpolator((xx, yy, zz), 
                                                      PSF, 
                                                      method = intmethod, 
                                                      bounds_error = bounds_error,
                                                      fill_value = extraolation_method)
            self.PSFs[key] = PSFdonut
        else:
            raise UserWarning(f"{phasepattern} PSF name not defined.")
        
        return key


    def beadsize(self, psf, R):
        if R == 0:
            return psf
        
        dr = self.parameters['addpar']['dr']
        dz = self.parameters['addpar']['dz']
        
        nx = np.arange(psf.shape[0])
        nx = (nx - np.mean(nx)) * dr
        
        ny = np.arange(psf.shape[1])
        ny = (ny - np.mean(ny)) * dr
        
        nz = np.arange(psf.shape[2])
        nz = (nz - np.mean(nz)) * dz
        
        X, Y, Z = np.meshgrid(nx, ny, nz)
        circpsf = np.double((X**2 + Y**2 + Z**2) <= R**2)
        circpsf /= np.sum(circpsf)
        
        psfo = convnfft(psf, circpsf)
        
        return psfo
    
    def setpinhole(self, lambda_nm=600, AU=1, diameter=None, offset=[0,0]):
        """
        Sets the pinhole size based on given parameters.

        Parameters
        ----------
        lambda_nm (float): Wavelength in nm (default: 600 nm)
        AU (float): Airy unit scaling factor (default: 1)
        diameter (float or None): Pinhole diameter in nm (default: computed if None)
        NA (float): Numerical aperture (default: 1.5)
        refractive_index (float): Medium refractive index (default: 1.5)
        """
        if len(offset) != 2:
            raise ValueError('PSF_vectorial.setpinhole: offset needs to have two entries [x,y]')
        
        # Compute diameter if not provided
        if diameter is None:
            diameter = np.round(AU*1.22*self.parameters['sys']['loem']*1e9/self.parameters['sys']['NA'])
            # print(f"calculated diameter is: {diameter}")

        phdefined = self.PSFph is not None and self.pinholepar is not None
        notchanged = phdefined and (diameter == self.pinholepar['diameter']) and offset == self.pinholepar['offset']
        calculated = notchanged and genkey('pinhole', [diameter,offset]) in self.PSFs.keys()

        if not calculated:
            self.PSFph = self.calculatePSFs('pinhole',
                                            Lxs = [diameter,offset],
                                            forcecalculation=True)

        self.pinholepar = {}
        self.pinholepar['offset'] = offset
        self.pinholepar['diameter'] = diameter

    def imagestack(self, key, show=False):
        if key not in self.PSFs.keys():
            key = self.calculatePSFs(key, 0)
        psf = self.PSFs[key]
        vout = psf.interp.values

        if self.PSFph is not None:
            # Apply pinhole
            phpsf = self.PSFs[self.PSFph]
            vout = vout*phpsf.interp.values

        return vout.transpose(1,0,2)
    
    def normfactor(self, phasemask, number):
        return self.PSFs[f"{phasemask}{number}"].normalization

File: psfs/PsfVectorial.m
classdef PsfVectorial<Psf
    properties
        parameters
        PSFs=dictionary;
        PSFph=[];
        normfactgauss
        pinholepar
        sigma
        fwhm
    end
    methods
        function obj=PsfVectorial(varargin)
            obj@Psf(varargin{:});
            addpath([fileparts(mfilename('fullpath')) '/PSF_simulation/library']);
            obj.parameters=obj.loadparameters('/settings/defaultsystemparameters_vectorialPSF.m');
        end
        function setpar(obj,varargin)
            if nargin >2 %structure passed on
                spar=struct(varargin{:});
            else
                spar=varargin{1};
            end
            % overwrite any of the parameter fields if they exist,
            % otherwise add to addpar
            fn=fieldnames(spar);
            fnpar=fieldnames(obj.parameters);
            recalculate=false;
            for k=1:length(fn)
                assigned=false;
                for l=1:length(fnpar)
                    if isfield(obj.parameters.(fnpar{l}),fn{k})
                        if ~isequal(obj.parameters.(fnpar{l}).(fn{k}),spar.(fn{k}))
                            obj.parameters.(fnpar{l}).(fn{k})=spar.(fn{k});
                            recalculate=true;
                        end
                        assigned=true;
                    end
                end
                if ~assigned
                    obj.parameters.addpar.(fn{k})=spar.(fn{k});
                    % recalculate=true;
                end 
            end
            if recalculate 
                obj.PSFs=dictionary;%delete pre-calculated PSFs
                obj.normfactgauss=[];
                if ~isempty(obj.pinholepar)
                    pp=struct2pairs(obj.pinholepar);
                    obj.setpinhole(pp{:})
                end
            end
        end
        function [idet,phfac]=intensity(obj, flpos ,patternpos, phasepattern, L)
           
            if nargin<5
                key=phasepattern;
            else
                key=phasepattern+L;
            end
            %flpos with respect to optical axis ([0,0] of PSF coordinate
            %system)
            % patternpos: position by EOD only of excitation pattern, no
            % descanning. Pinhole is with respect to [0,0] (but can be
            % shifted in pinhole definition).
            
            psfint=obj.PSFs(key);
            phkey=obj.PSFph;
            flposrel=flpos-patternpos;
            iexc=psfint.interp(flposrel)+obj.zerooffset; 
            iexc=max(iexc,0);
            if isempty(phkey) 
                phfac=ones(size(iexc));
            else
                psfph=obj.PSFs(phkey);
                phfac=psfph.interp(flpos);
            end
            idet=iexc.*phfac;
        end
        function key=calculatePSFs(obj,phasepattern,Lxs,forcecalculation)
            if nargin<4
                forcecalculation=false;
            end
            if isnumeric(Lxs)
                Lx=Lxs;
                Lxs=num2str(Lxs);
            else
                Lx=str2num(Lxs);
            end
            key=string(phasepattern)+Lxs;
            if obj.PSFs.isConfigured && obj.PSFs.isKey(key) && ~forcecalculation
                return
            end
            fprintf(key + ", ")
            intmethod='cubic'; %linear,cubic?
            extraolation_method='nearest'; 

            opt=obj.parameters.opt;
            out=obj.parameters.addpar;
             sys=obj.parameters.sys;
            % interpolant:
            % nx=(-opt.radiusCanvas:out.dr:opt.radiusCanvas)*1e9; %to nm
            % nz=(-opt.depthCanvas:out.dz:opt.depthCanvas)*1e9; %to nm
            % [X,Y,Z] = ndgrid(nx,nx,nz);
            
           
            % sigma from wavelength, NA
            fwhm=0.51*sys.loem/sys.NA*1e9;
            obj.sigma=fwhm/2.35;
            
            if isempty(obj.normfactgauss)
                sys.Ei = {'circular'};
                outc=effField(sys,out, opt);      
                outc=effIntensity(sys,outc);
                zmid=ceil(size(outc.I,4)/2);
                % normfact=sum(sum(outc.I(1,:,:,zmid)));%normalized to integral =1
                normfact=max(max(outc.I(1,:,:,zmid)));%normalized to max =1
                obj.normfactgauss=normfact;
            else
                normfact=obj.normfactgauss;
            end
                
            % %fit Gaussian:
            % vg=squeeze(outc.I(1,:,ceil(end/2),zmid))';
            % x=1:length(vg); x=x'-mean(x); x=x*out.dr*1e9;
            % fitp=fit(x,vg,'gauss1');
            % obj.sigma=fitp.c1/sqrt(2);

            switch phasepattern
                case 'flat'
                    sys.Ei = {'circular'};    
                    sys=addzernikeaberrations(sys,out);
                    out=effField(sys,out, opt);      
                    out=effIntensity(sys,out);
                    PSF=squeeze(out.I(:,:,:,:))/normfact;
                    PSF = obj.beadsize(PSF, sys.beadradius);
                    [PSF,PSFdonut.normalization]=normpsf(PSF);
                    [X,Y,Z]=meshgrid4PSF(PSF,out.dr,out.dz);
                    PSFdonut.interp = griddedInterpolant(X,Y,Z,PSF,intmethod, extraolation_method);              
                    obj.PSFs(key)=PSFdonut;
                case {'halfmoonx','halfmoony'}
                    %convert L into phase
                    dxdphi=1.4; %nm/degree, from LSA paper Deguchi
                    % dzdphi=-3.6; %nm/degree
                    sys.delshift = deg2rad(Lx/dxdphi);
                    sys.Ei = {'halfmoon', 'linear'};  
                    sys=addzernikeaberrations(sys,out);
                    out=effField(sys,out, opt);      
                    out=effIntensity(sys,out);
                    PSF=squeeze(out.I(:,:,:,:))/normfact;
                    PSF = obj.beadsize(PSF, sys.beadradius);
                    [PSF,normalization]=normpsf(PSF);
                    [X,Y,Z]=meshgrid4PSF(PSF,out.dr,out.dz);
                    PSFx.interp = griddedInterpolant(X,Y,Z,permute(PSF,[2,1,3]),intmethod, extraolation_method);
                    PSFy.interp = griddedInterpolant(X,Y,Z,PSF,intmethod, extraolation_method);     
                    PSFx.normalization=normalization;
                    PSFy.normalization=normalization;
                    obj.PSFs("halfmoonx"+Lxs) =PSFx; 
                    obj.PSFs("halfmoony"+Lxs)=PSFy; 
                case 'pinhole' %here
                    sys.Ei = {'circular'};
                    sys=addzernikeaberrations(sys,out);
                    phdiameter=Lx(1);
                    if length(Lx)==3
                        phpos=[Lx(2) Lx(3)];
                    else
                        phpos=[0 0];
                    end
                    out=effField(sys,out, opt);      
                    out=effIntensity(sys,out);
                    psfg=squeeze(out.I);
                    % psfgmaxnorm=psfg/normfact;
                    norm=sum(sum(psfg(:,:,round(end/2))));
                    psfg=psfg/norm;
                    pixelsize=opt.pixSize*1e9;
                    nx=1:size(psfg,1);
                    nx=(nx-mean(nx))*pixelsize;
                    [Xk,Yk]=meshgrid(nx);
                    kernel=double((Xk-phpos(1)).^2+(Yk-phpos(2)).^2<(phdiameter/2)^2);
                    psfph=conv2fft(psfg,kernel);
                    [psfph,PSFdonut.normalization]=normpsf(psfph);
                    [X,Y,Z]=meshgrid4PSF(psfph,out.dr,out.dz);
                    PSFdonut.interp = griddedInterpolant(X,Y,Z,psfph,intmethod,extraolation_method);
                    obj.PSFs(key)=PSFdonut;   
                case 'tophat'
                    dzdphi=-3.6; %nm/degree
                    sys.delshift = deg2rad(Lx/dzdphi);
                    sys.Ei = {'pishift', 'circular'};  
                    sys=addzernikeaberrations(sys,out);
                    out=effField(sys,out, opt);      
                    out=effIntensity(sys,out);
                    PSF=squeeze(out.I(:,:,:,:))/normfact;
                    PSF = obj.beadsize(PSF, sys.beadradius);
                    [PSF,PSFdonut.normalization]=normpsf(PSF);
                    [X,Y,Z]=meshgrid4PSF(PSF,out.dr,out.dz);
                    PSFdonut.interp = griddedInterpolant(X,Y,Z,PSF,intmethod, extraolation_method);
                    obj.PSFs(key)=PSFdonut;   
                case 'vortex'
                    sys.Ei = { 'phaseramp',  'circular'};  
                    sys=addzernikeaberrations(sys,out);
                    out=effField(sys,out, opt);      
                    out=effIntensity(sys,out);
                    PSF=squeeze(out.I(:,:,:,:))/normfact;
                    PSF = obj.beadsize(PSF, sys.beadradius);
                    [PSF,PSFdonut.normalization]=normpsf(PSF);
                    [X,Y,Z]=meshgrid4PSF(PSF,out.dr,out.dz);
                    PSFdonut.interp = griddedInterpolant(X,Y,Z,PSF,intmethod, extraolation_method);              
                    obj.PSFs(key)=PSFdonut;
                otherwise
                    warning(phasepattern+" PSF name not defined")
            end
        end
        function psfo=beadsize(obj,psf,R)
            if R==0
                psfo=psf;
                return
            end
            dr=obj.parameters.addpar.dr; dz=obj.parameters.addpar.dz;
            nx=1:size(psf,1);
            nx=(nx-mean(nx))*dr;
            ny=1:size(psf,2);
            ny=(ny-mean(ny))*dr;
            nz=1:size(psf,3);
            nz=(nz-mean(nz))*dz;
            [X,Y,Z]=meshgrid(nx,ny,nz);
            circpsf=double((X).^2+(Y).^2+(Z).^2<=R^2);
            circpsf=circpsf/sum(circpsf(:));
            psfo=convnfft(psf,circpsf);
            % psfo=ifftshift(ifftshift(ifftshift(psfo,1),2),3); %inconsitent use of fftshift, but ok because bead is symmetric
        end
        function setpinhole(obj, args)
            arguments
                obj 
                args.lambda=600; %nm
                args.AU =1; %in airy units
                args.diameter=[] %nm
                args.offset=[0,0];
            end
            if length(args.offset)~=2
                error('PSF_vectorial.setpinhole: args.offset needs to have two entries (x,y)')
            end
            if isempty(args.diameter)
                % [sys,opt,out]=systemparameters;
                args.diameter=round(args.AU*1.22*obj.parameters.sys.loem*1e9/obj.parameters.sys.NA); %single nm accuracy should be sufficient
            end

            %sdetermine if PH is already calculated
            phdefined=~isempty(obj.PSFph) && ~isempty(obj.pinholepar);
            notchanged=phdefined && (args.diameter == obj.pinholepar.diameter...
                    && all(args.offset == obj.pinholepar.offset)); 

            calculated=notchanged && ((obj.PSFs.isConfigured) && isKey(obj.PSFs,"pinhole"+num2str([args.diameter,args.offset])));

            % if isempty(obj.PSFph)||(~isempty(obj.pinholepar) && (args.diameter ~= obj.pinholepar.diameter...
            %         || any(args.offset ~= obj.pinholepar.offset)) ) || (~isempty(obj.PSFs) && ~isKey(obj.PSFs,"pinhole"+num2str([args.diameter,args.offset])))
            if ~calculated
                obj.PSFph=obj.calculatePSFs('pinhole',[args.diameter,args.offset],1);
            end
            obj.pinholepar.offset=args.offset;
            obj.pinholepar.diameter=args.diameter;
        end
       
        function savePSF(obj,name)
            PSFinterpolant=obj.PSFinterpolant;
            save(name,'PSFinterpolant')
        end
        function loadPSF(obj,name)
            load(name,'PSFinterpolant');
            obj.PSFinterpolant=PSFinterpolant;
        end
        function plotprofile(obj,key)
            psf=obj.PSFs(key);
            inten=psf.Values;
            xaxv=(psf.GridVectors{1});
            figure(33)
            plot(xaxv, inten(:,round(end/2),round(end/2)))
        end
        function showpsf(obj,key)
            psf=obj.PSFs(key);
            inten=psf.Values;
            if isempty(obj.PSFph)
                psfph=1;
            else
                psfphx=obj.PSFs(obj.PSFph);
                psfph=psfphx.Values;
            end
            psftot=inten.*psfph;
            imx(psftot);
        end
        function par=loadparameters(obj,filen)
            [fpath,fname,ext]=fileparts(filen);
            switch ext
                case '.m'
                    if exist(fpath,"dir")
                        addpath(fpath)
                    end
                    [sys,opt,out]=eval(fname);
                    [sys2,out]=effInit_oil_exc(sys,out,opt);
                    sys=copyfields(sys,sys2);
                    par.sys=sys;
                    par.opt=opt;
                    par.addpar=out;
                case '.json'
                    disp('json not implemented yet')
            end
        end
        % function out=fwhm(obj)
        %      out=obj.sigma*2.35*1.2; %comes from comparison with simple donut
        % end
        function loadexperimental(obj,key,fname)
            [~,fnh,ext]=fileparts(fname);
            if contains(fname,"_3dcal") && ext==".mat" %3Dcal.mat
                l=load(fname);
                PSFvol=l.SXY(1).PSF{1};
                px=l.parameters.pixelsize{1}(1)*1000; %nm
                pz=l.parameters.dz*1000;
                spsf=size(PSFvol);
                nx=1:spsf(1);nx=nx-mean(nx);nx=nx*px;
                ny=1:spsf(2);ny=ny-mean(ny);ny=ny*px;
                nz=1:spsf(3);nz=nz-mean(nz);nz=nz*pz;
                [X,Y,Z] = ndgrid(nx,nx,nz);
                intmethod='cubic'; %linear,cubic?
                PSFexp = griddedInterpolant(X,Y,Z,PSFvol,intmethod);              
                obj.PSFs(key+"0")=PSFexp;
            

            else
                disp('not implemented')
            end
            %uiPSF
            %tiffstack
        end
        function vout=imagestack(obj,key,args)
            arguments
                obj
                key
                args.show=false;
            end
            if ~isKey(obj.PSFs,key)
                key=key+"0";
            end
            psf=obj.PSFs(key);
            vout=psf.interp.Values;
            if ~isempty(obj.PSFph)
                phpsf=obj.PSFs(obj.PSFph);
                vout=vout.*phpsf.interp.Values  ; 
            end

            if args.show
                imx(double(vout))
            end
        end
        function nf=normfactor(obj,phasemask,number)
            nf=obj.PSFs(string(phasemask)+number).normalization;
        end
    end
end

function sys=addzernikeaberrations(sys,addpar)
if isfield(sys,'Zr') && ~isempty(sys.Zr)
    sys.Ei{end+1}='zernike';
elseif isfield(addpar,'Zr') && ~isempty(addpar.Zr)
    sys.Ei{end+1}='zernike';
    sys.Zr=addpar.Zr;
end
end


function [PSF,normf]=normpsf(PSF)
normf=max(PSF(:));
PSF=PSF/normf;
end

function [X,Y,Z]=meshgrid4PSF(PSF,dr,dz)
nx=((1:size(PSF,1))-mean((1:size(PSF,1))))*dr*1e9;
nz=((1:size(PSF,3))-mean((1:size(PSF,3))))*dz*1e9;
[X,Y,Z]=ndgrid(nx,nx,nz);
end